server/src/main/java/org/elasticsearch/repositories/RepositoryData.java

```diff
@@ -958,9 +958,7 @@ private static IndexVersion parseIndexVersion(XContentParser.Token token, XConte
             return IndexVersion.fromId(parser.intValue());
         } else {
             XContentParserUtils.ensureExpectedToken(XContentParser.Token.VALUE_STRING, token, parser);
-            Version v = Version.fromString(parser.text());
-            assert v.before(Version.V_8_10_0);
-            return IndexVersion.fromId(v.id);
+            return IndexVersion.fromId(Version.fromString(parser.text()).id);
```

Comment:
Do we want to assert that this is an `IndexVersion` that has a corresponding `Version`? ie is not a detached version? We don't want to create an IndexVersion for 8.12.0 when it should in fact be 8,500,002 for that release version

Reply:
Would that actually crash the node, or just make the repository unusable? If there is a bit flip in the index version it seems like we would want to not trust anything in this metadata (which separately begs the question, should we have some integrity check on the data like a checksum?).

Reply:
If running with assertions enabled it'd crash the node. If running with assertions disabled then kinda nothing would happen, we might just skip over this snapshot in some situations. But still IMO troublesome to have an assertion about the contents of a file that could be changed by external forces.

And yes the lack of integrity checks on this file is something we've noted before. Hard to fix without a breaking change tho.

Tests for integrity checks in particular are hard to write if the deserialization code asserts things that might not be true on a corrupt file.

Reply:
Can we have some sort of indication this might inject an incorrect `IndexVersion` into the system? A log warning, to help track down following errors if they occur?

Every other time I'm going from Version -> Index/TransportVersion I've enforced that it's only a version with a valid corresponding index/transport version, as documentation for future devs if anything else.

Reply:
Sure, an error log seems more graceful. See 6559b6c46200141f966365fab90f5e2187b31cc1.